src/api_rework.jl

```diff
@@ -0,0 +1,194 @@
+# I am putting everything in here for now. The contents of this file should be moved around in the future.
+
+"""
+    peakproms!(pks) --> NamedTuple
+    peakproms!() --> Function
+
+# Optional keyword arguments
+- `min`: Filter out any peak with a height smaller than `min`.
+- `max`: Filter out any peak with a height greater than `min`.
+- `strict`: How to handle `NaN` and `missing` values. See documentation for more details. Default to `true`.
+
+Find the prominences of the peaks in `pks`, and filter out any peak 
+with a prominence smaller than `min` or greater than `max`.
+The prominences are returned in the field `:proms` of the returned named tuple.
+
+If the positional argument `pks` is omitted, a function is returned such
+that `peakproms!(pks)` is equivalent to `pks |> peakproms!`.
+
+Note: This function mutates the vectors stored in the NamedTuple `pks`, 
+and not the named tuple itself.
+
+See also: [`peakwidths!`](@ref), [`peakheights!`](@ref)
+
+# Examples
+```jldoctest
+julia> data = [1, 5, 1, 3, 2];
+
+julia> pks = findmaxima(data);
+
+julia> pks = peakproms!(pks)
+(indices = [2, 4], heights = [5, 3], data = [1, 5, 1, 3, 2], proms = Union{Missing, Int64}[4, 1])
+
+julia> data |> findmaxima |> peakproms!
+(indices = [2, 4], heights = [5, 3], data = [1, 5, 1, 3, 2], proms = Union{Missing, Int64}[4, 1])
+```
+"""
+function peakproms!(pks::NamedTuple; minprom=nothing, maxprom=nothing, min=minprom, max=maxprom, strict=true)
+    if !hasproperty(pks, :proms)
+        # Avoid filtering by min/max/strict here, so that it always happens outside if-statement.
+        # Pro: one less edge case. Con: More internal allocations
+        _, proms = peakproms(pks.indices, pks.data; strict)
+        pks = merge(pks, (; proms))
+    end
+    filterpeaks!(pks, min, max, :proms)
+    return pks
+end
+peakproms!(; kwargs...) = pks -> peakproms!(pks; kwargs...)
+
+"""
+    peakproms(pks) --> NamedTuple
+    peakproms() --> Function
+
+# Optional keyword arguments
+- `min`: Filter out any peak with a height smaller than `min`.
+- `max`: Filter out any peak with a height greater than `min`.
+- `strict`: How to handle `NaN` and `missing` values. See documentation for more details. Default to `true`.
+
+Non-mutation version of `peakproms!`. Note that 
+this copies all vectors in `pks`, meaning that 
+it is less performant. See the docstring for 
+`peakproms!` for more information.
+"""
+peakproms(pks::NamedTuple; kwargs...) = peakproms!(deepcopy(pks); kwargs...)
```

Comment:
That is  good point. I guess copying everything is more consistent and simple, but not copying data is more preformant. I am leaning torwards the former, because if you are using this method, performance is not what matters most to you anyways, and the extra risk of bugs that come from users that did not catch that the data was not copied is not worth it. So while you make a good point that the copy is completely wasteful, I suggest we do it that way anyways to avoid unexpected bugs.

Reply:
I did as I said. Can I resolve this?